### ISSUE TITLE
CPixelView doesn't handle reversed direction and color utility functions

### ISSUE THREAD

**Issue by MFornander** [FastLED/FastLED#481]
We just discovered a few bugs in the rainbow and gradient methods inside CPixelView when the view is mapped on a reverse direction led array.  We fixed the rainbow but I'm filing this issue to remember to fix all methods and perhaps even add a few perlin noise methods on CPixelView.

Repro steps
1. Fold one physical strip of 100 LEDs in half and connect to driver.
2. Add all 100 LEDs using addLeds.
3. Create two "logical" view of LEDs using CPixelView's ctor, one with (leds, 0,49) straight and the other as (leds, 99,50) reversed direction.
4. Call fill_rainbow twice on both views with the same values.

Expected result:
Both logical strips showing the same rainbow, with identical colors side by side, as if they were two different strips connected in parallel to controller.

Actual result:
Incorrectly mapped rainbow color on the reverse direction view.

I'll file a Merge Request with fixes sometime after the dust settles this year.

**Comment by MFornander** [comment]
Note to self since we had to recreate the fix again today:

pixelset.h @ 166
OLD AND BUSTED
`::fill_rainbow(leds+len+1,-len,initialhue,deltahue);`

NEW HOTNESS
`::fill_rainbow(leds+len+1,-len,initialhue-deltahue*len,-deltahue);`

Still planning to file that MR but want to look at the sibling gradient methods to make sure they are not flawed as well.

**Comment by MFornander** [comment]
The reverse direction was absolutely awesome for us.  We have two strands folded on a pole, and thus have four CPixelViews as if we had four physical strands.  This saved us connectors and development time since other incarnations are not folded.  Very cool abstraction.  Don't gut it :)

**Comment by focalintent** [comment]
Sorry - should've been clearer - I'm going to gut the implementation - I've been meaning to look at some different ways of doing range selectors as a general idea, to give something a lot more flexible (forward ranges, backward ranges, repeat patterns, skip lists, etc...)

**Comment by kriegsman** [comment]
Cleaning up old issues; closing this one, since @focalintent is no longer around to work on it.  May reopen it in the future.

**Comment by MFornander** [comment]
Hey @kriegsman would you like an MR that addresses this in the current codebase?

**Comment by kriegsman** [comment]
That would be great actually, if you can and feel like it.

**Comment by zackees** [comment]
Deferred by automated triage: Investigated in PR #3804, which adds tests pinning reverse-direction behavior. The originally-reported fill_rainbow bug was fixed in #1756 and is now verified exact, as is fill_solid; the 2-color gradients differ only by sub-step rounding (<=3/255). Remaining is a real defect: fill_gradient with 3 or 4 colors places the middle color one pixel off under reversal (~22/255), because fl/gfx/fill.h pivots at floor(numLeds/2), which is not reversal-symmetric. Fixing needs either a symmetric pivot in fill.h (moves output for every caller) or fill-forward-then-reverse-in-place in CPixelView (changes the cost model) -- a design call, not a one-expression fix. Also still open: perlin-noise methods on CPixelView, and focalintent's plan to replace range selectors wholesale.
